### PR TITLE
[megatron] feat: checkpoint save as HF PEFT format

### PR DESCRIPTION
**Pull request by HollowMan6**
### What does this PR do?

New API endpoints added on Megatron-Bridge side, need https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/2574 (has been merged)

### Checklist Before Starting

- [X] Search for similar PRs. Paste at least one query link here: ...
- [X] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `veomni`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`, `fully_async`, `one_step_off`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluation results, etc.

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [X] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [X] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [X] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs).
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [X] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)
- [ ] If your PR is related to the `recipe` submodule, please also update the reference to the submodule commit via `git submodule update --remote` or `cd recipe && git pull origin main`.

### CODE REVIEW

File: verl/utils/checkpoint/megatron_checkpoint_manager.py [hunk 1]
```diff
@@ -535,11 +526,15 @@ def save_checkpoint(self, local_path: str, hdfs_path: str = None, global_step: i
...
-                generate_model=False,
+                generate_model=self.should_save_model and self.peft_cls is not None,
                 generate_optimizer=self.should_save_optimizer,
                 generate_extra=self.should_save_extra,
                 metadata=sharded_sd_metadata,
             )
+            if self.peft_cls is not None:
+                from megatron.bridge.training.checkpointing import apply_peft_adapter_filter_to_state_dict
+
+                state_dict = apply_peft_adapter_filter_to_state_dict(state_dict, self.peft_cls)
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply verl/utils/checkpoint/megatron_checkpoint_manager.py]: fixed

File: verl/utils/megatron_utils.py [hunk 2]
```diff
@@ -235,7 +240,12 @@ def peft_pre_wrap_hook(model):
                     adapter_path = getattr(peft_config, "adapter_path", None)
                     if adapter_path is not None and adapter_path:
                         print(f"Loading adapter weights from: {adapter_path}")
-                        load_adapter_checkpoint(transformed_model, adapter_path)
+                        sharded_state_dict = _generate_model_state_dict(transformed_model, {})
+                        sharded_state_dict = apply_peft_adapter_filter_to_state_dict(sharded_state_dict, peft_cls)
+                        loaded_state_dict = dist_checkpointing.load(sharded_state_dict, str(adapter_path))
+                        for vpp_rank, model_chunk in enumerate(transformed_model):
+                            model_key = "model" if len(transformed_model) == 1 else f"model{vpp_rank}"
+                            model_chunk.load_state_dict(loaded_state_dict[model_key], strict=False)
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply verl/utils/megatron_utils.py]: fixed

File: verl/utils/checkpoint/megatron_checkpoint_manager.py [hunk 3]
```diff
@@ -392,13 +392,18 @@ def load_checkpoint(self, local_path: str, hdfs_path: str = None, del_local_afte
...
             self.should_load_optimizer,
             self.should_load_extra,
             is_loading=True,
             metadata=sharded_sd_metadata,
         )
+        if self.peft_cls is not None:
+            from megatron.bridge.training.checkpointing import apply_peft_adapter_filter_to_state_dict
+
+            sharded_state_dict = apply_peft_adapter_filter_to_state_dict(sharded_state_dict, self.peft_cls)
         log_with_rank(f"Generated state dict for loading: {sharded_state_dict.keys()}", rank=self.rank, logger=logger)
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply verl/utils/checkpoint/megatron_checkpoint_manager.py:446]: fixed

File: verl/utils/checkpoint/megatron_checkpoint_manager.py [hunk 4]
```diff
@@ -535,11 +526,15 @@ def save_checkpoint(self, local_path: str, hdfs_path: str = None, global_step: i
             # Generate optimizer and exra state dicts
             sharded_sd_metadata = self._build_sharded_state_dict_metadata()
             state_dict = self.generate_state_dict(
-                generate_model=False,
+                generate_model=self.should_save_model and self.peft_cls is not None,
                 generate_optimizer=self.should_save_optimizer,
                 generate_extra=self.should_save_extra,
                 metadata=sharded_sd_metadata,
             )
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply verl/utils/checkpoint/megatron_checkpoint_manager.py:569]: fixed

File: verl/utils/checkpoint/megatron_checkpoint_manager.py [hunk 5]
```diff
@@ -588,7 +564,17 @@ def save_checkpoint(self, local_path: str, hdfs_path: str = None, global_step: i
...
-                    self.bridge.save_hf_weights(self.model, hf_ckpt_path)
+                    if self.peft_cls is not None:
+                        hf_adapter_ckpt_path = os.path.join(hf_ckpt_path, "adapter")
+                        self.bridge.save_hf_adapter(self.model, hf_adapter_ckpt_path, self.peft_cls)
+                        log_with_rank(
+                            f"Saved HF PEFT adapter checkpoint to {hf_adapter_ckpt_path}",
+                            rank=self.rank,
+                            logger=logger,
+                            log_only_rank_0=True,
+                        )
```
[the remark this answers is not in the document]
Comment (HollowMan6) [reply verl/utils/checkpoint/megatron_checkpoint_manager.py:608]: fixed

**Comment by HollowMan6** [comment]
Thank you!

> Can I understand this as runnable PEFT format saving refactor?

Yes, mainly to move away from Verl's own customized PEFT checkpointing format (was introduced by me in https://github.com/verl-project/verl/pull/4063 hh) into the official APIs introduced by https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/2574

> The saving methods before have some bugs?

I don't find bugs previously unless other people find anything, the main goal was to centralize the code to Megatron-Bridge and keep related Verl codebase clean.

> Do we need to modify some docs in another PR?

The PEFT checkpointing format stuff was not documented in https://github.com/verl-project/verl/blob/main/docs/advance/ppo_lora.rst, so feel free to add a section about it.
